homeassistant/components/broadlink/remote.py

```diff
@@ -0,0 +1,418 @@
+
+        code = b64encode(code).decode("utf8")
+        return (
+            code
+            if not toggle
+            else [code, await self._async_capture_code(command, timeout=timeout)]
+        )
+
+    async def _async_attempt(self, function, *args):
+        """Retry a socket-related function until it succeeds."""
+        for retry in range(DEFAULT_RETRY):
+            if retry and not await self._async_connect():
+                continue
+            try:
+                await self.hass.async_add_executor_job(function, *args)
+            except socket.error:
+                continue
+            return
+        raise ConnectionError
+
+    async def _async_connect(self):
+        """Connect to the remote."""
+        try:
+            auth = await self.hass.async_add_executor_job(self._api.auth)
+        except socket.error:
+            auth = False
+        if auth and not self._available:
+            _LOGGER.warning("Connected to the remote")
+            self._available = True
+        elif not auth and self._available:
+            _LOGGER.warning("Disconnected from the remote")
+            self._available = False
+        return auth
+
+    async def async_retrieve_data(self):
+        """Load dictionary of commands and toggle data from files."""
+        try:
+            await self._async_load_commands(self._command_file)
+            await self._async_load_toggle_data(self._toggle_file)
+        except FileNotFoundError:
+            pass  # No problem. The file will be created soon.
+        except (IOError, json.JSONDecodeError, vol.MultipleInvalid):
+            return False
+        return True
+
+    async def _async_load_commands(self, file_path):
+        """Load dictionary of commands from a JSON file."""
+        try:
+            with open(file_path, "r") as json_file:
+                self._commands = COMMAND_SCHEMA(json.load(json_file))
+        except FileNotFoundError:
+            raise
+        except IOError as err:
+            _LOGGER.error("Failed to load '%s': %s", file_path, err.strerror)
+            raise
+        except json.JSONDecodeError as err:
+            _LOGGER.error("Failed to load '%s': %s", file_path, err)
+            raise
+        except vol.MultipleInvalid as err:
+            _LOGGER.error("Failed to load '%s': %s", file_path, err)
+            raise
+
+    async def _async_store_commands(self, file_path):
+        """Dump dictionary of commands into a JSON file."""
+        try:
+            with open(file_path, "w") as json_file:
+                json.dump(self._commands, json_file, indent=4)
+        except IOError as err:
+            _LOGGER.error("Failed to write '%s': %s", file_path, err.strerror)
+
+    async def _async_load_toggle_data(self, file_path):
+        """Load dictionary of toggle values from a binary file.
+
+        You should call ``self._async_load_commands()`` first.
+        """
+        try:
+            with open(file_path, "rb") as bin_file:
```

Comment:
Haven't looked at the other parts of the PR, but here you're doing I/O inside the event loop (async method), which is not allowed. 

Also, use the storage helper `homeassistant.helpers.storage.Store()` to do such things.

Reply:
I'm sorry, I didn't know about the existence of this helper class. I will refactor the code to fit the standards.

Reply:
One more thing. What do you think would be the best way to store/update a boolean each time the user sends a toggle (eg. power) command? I need to store this information persistently so that the program knows which code (eg. on/off) to send even after a reboot.

I would personally adopt SQL, but I don't know if it's allowed, so I adopted the current solution of storing the booleans in a binary file, which seemed less costly in terms of I/O. Should I use storage.Store() instead?

Reply:
> I would personally adopt SQL, but I don't know if it's allowed, so I adopted the current solution of storing the booleans in a binary file, which seemed less costly in terms of I/O. Should I use storage.Store() instead?

It's a bad design. Only an Entity in home-assistant should have a state.

Reply:
`storage.Store` has a delay `async_delay_save`. Just set delay to 15 minutes. It will write after 15 minutes of no change or when Home Assistant shuts down.

Reply:
> > I would personally adopt SQL, but I don't know if it's allowed, so I adopted the current solution of storing the booleans in a binary file, which seemed less costly in terms of I/O. Should I use storage.Store() instead?
> 
> It's a bad design. Only an Entity in home-assistant should have a state.

I understand, but it's not a state, it's a flag. It indicates the necessity of sending an alternative command. In the example I used on/off but it could be the mute button. Some remotes operate with this toggle mechanism that sends different code with each click.. What this implementation does is mimic the process so that the transmission has a good response rate.

Reply:
> `storage.Store` has a delay `async_delay_save`. Just set delay to 15 minutes. It will write after 15 minutes of no change or when Home Assistant shuts down.

Ok. I'll do it.

Reply:
> I understand, but it's not a state, it's a flag. It indicates the necessity of sending an alternative command. 
What do you do when the flag becomes out of sync with the real thing?
In this case, I suggest using an external entity (switch, binary_sensor, etc.) to hold the assumed state or use Broadlink switch platform.

Reply:
> What do you do when the flag becomes out of sync with the real thing?
> In this case, I suggest using an external entity (switch, binary_sensor, etc.) to hold the assumed state or use Broadlink switch platform.

I've been using this integration for a few months and it never went out of sync. Apparently the code generated by the original remote varies in a non-binary way, which helps a lot.
If for some reason this goes out of sync, the most that can go wrong is a single repeated command being sent and ignored by the device, which would be the natural behavior without monitoring this flag. And in the next command, everything is working again.
So I'm going to take @balloob's suggestion and use the storage method, which seems more appropriate for internal details that don't need to be visible to the end user.